File: clients/phplib/Conjecture/Text.php
<?php

// This is bascially an exact replica of com.etsy.conjecture.text.Text
class Conjecture_Text {

    private $input = null;

    static function build($text) {
        return new Conjecture_Text($text);
    }

    function __construct($text) {
        $this->input = $text;
    }

    function toString() {
        return $this->input;
    }

    function replaceNumbers($replacement = "_num_") {
        $text = preg_replace("/[0-9]+/", $replacement, $this->input);
        return new Conjecture_Text(preg_replace("/".$replacement."\\s+".$replacement."/", $replacement, $text));
    }

    function replaceHTMLEscapes($replacement = " ") {
        return new Conjecture_Text(preg_replace("/&[^;]+;/", $replacement, $this->input));
    }

    function removeHTMLTags() {
        return new Conjecture_Text(preg_replace("/<.*?>/", " ", $this->input));
    }

    function replaceHTMLTags($replacement = " ") {
        return new Conjecture_Text(preg_replace("/<[^>]+>/", " ", $this->input));
    }

    function replaceNonAlphaNumeric($replacement = " ") {
        return new Conjecture_Text(preg_replace("/[^a-zA-Z0-9\\.\\s\\-]+/", $replacement, $this->input));
    }

    function replaceNonAlphaNumericUnderscore($replacement = " ") {
        return new Conjecture_Text(preg_replace("/[^a-zA-Z0-9\\.\\s\\-_]+/", $replacement, $this->input));
    }

    function replaceNonAlpha($replacement = " ") {
        return new Conjecture_Text(preg_replace("/[^a-zA-Z]+/", $replacement, $this->input));
    }

    function collapseHyphens() {
        return new Conjecture_Text(preg_replace("/--+/", "--", $this->input));
    }

    function collapseUnderscores() {
        return new Conjecture_Text(preg_replace("/__+/", "__", $this->input));
    }

    function collapsePeriods() {
        return new Conjecture_Text(preg_replace("/\.\.+/", "..", $this->input));
    }

    function stripPunctuation() {
        $temp = preg_replace("^[^A-Za-z0-9]", "", $this->input);
        return new Conjecture_Text(preg_replace("[^A-Za-z0-9]$", "", $temp));
    }

    // compact any white space
    function collapse() {
        return new Conjecture_Text(preg_replace("/\\s+/", " ", $this->input));
    }

    // remove any whitespace from the right of a string
    function rstrip() {
        return new Conjecture_Text(preg_replace("/\\s+$/", "", $this->input));
    }

    // remove any whitespace from the left of a string
    function lstrip() {
        return new Conjecture_Text(preg_replace("/^\\s+/", "", $this->input));
    }

    // remove any leading or trailing whitespace
    function strip() {
        return $this->rstrip()->lstrip();
    }

    // clean up any whitespace
    function wsclean() {
        return $this->strip()->collapse();
    }

    // remove any unprintable non-ASCII characters
    function removeUnprintables() {
        return new Conjecture_Text(preg_replace("/[^\\x20-\\x7E]/", "", $this->input));
    }

    function collapseWhitespaceAndPunc() {
        $text = $this->collapse()->collapseHyphens();
        return new Conjecture_Text(preg_replace("/\\.\\.+/", ".", $text->toString()));
    }

    function toLowerCase() {
        return new Conjecture_Text(strtolower($this->input));
    }

    function standardTextFilter() {
        return $this->removeHTMLTags()
                    ->replaceHTMLEscapes()
                    ->replaceNumbers()
                    ->replaceNonAlphaNumericUnderscore()
                    ->collapseHyphens()
                    ->collapseUnderscores()
                    ->wsclean();
    }

    function toArrayFromShingles($n) {
        $shingles = array();

        $chars = str_split($this->input);
        for ($i = 0; $i < count($chars) - $n + 1; $i++) {
            $shingle = array_slice($chars, $i, $n);
            $shingles[] = implode("", $shingle);
        }

        return $shingles;
    }

    function toSequenceFromShingles($n) {
        return new Conjecture_TextSequence($this->toArrayFromShingles($n));
    }
}

File: src/main/scala/com/etsy/conjecture/text/Text.scala
package com.etsy.conjecture.text

case class Text(val input: String) {

    private implicit def text2str(txt: Text): String = txt.input
    private implicit def str2text(str: String): Text = new Text(str)

    override def toString = input.toString

    def replaceNumbers(replacement: String = "_num_") = Text(input.replaceAll("[0-9]+", replacement).replaceAll(replacement + "\\s+" + replacement, replacement))

    def replaceHTMLEscapes(replacement: String = " ") = Text(input.replaceAll("&[^;]+;", replacement))

    def removeHTMLTags() = Text(input.replaceAll("<.*?>", " ")) //Text(XML.loadString(input).text)

    def replaceHTMLTags(replacement: String = " ") = Text(input.replaceAll("<[^>]+>", " "))

    def replaceNonAlphaNumeric(replacement: String = " ") = Text(input.replaceAll("[^a-zA-Z0-9\\.\\s\\-]+", replacement))

    def replaceNonAlphaNumericUnderscore(replacement: String = " ") = Text(input.replaceAll("[^a-zA-Z0-9\\.\\s\\-_]+", replacement))

    def replaceNonAlpha(replacement: String = " ") = Text(input.replaceAll("[^a-zA-Z]+", replacement))

    def collapseHyphens() = Text(input.replaceAll("--+", "--"))

    def collapseUnderscores() = Text(input.replaceAll("__+", "__"))

    def collapsePeriods() = Text(input.replaceAll("\\.\\.+", ".."))

    def toLowerCase() = Text(input.toLowerCase)

    def toUpperCase() = Text(input.toUpperCase)

    def stripPunctuation() = Text(input.replaceAll("^[^A-Za-z0-0]+", "").replaceAll("[^A-Za-z0-9]+$", ""))

    // compact any white space
    def collapse() = Text(input.replaceAll("\\s+", " "))

    // remove any whitespace from the right of a string
    def rstrip() = Text(input.replaceAll("\\s+$", ""))

    // remove any whitespace from the left of a string
    def lstrip() = Text(input.replaceAll("^\\s+", ""))

    // remove any leading or trailing whitespace
    def strip() = Text(input.trim)

    // clean up any whitespace
    def wsclean() = strip().collapse()

    // remove any unprintable non-ASCII characters
    def removeUnprintables(input: String) = Text(input.replaceAll("[^\\x20-\\x7E]", ""))

    def collapseWhitespaceAndPunc = Text(input.replaceAll("\\s+", " ")
        .replaceAll("[\\-]+", "-")
        .replaceAll("[\\.]+", "."))

    def standardTextFilter = Text(removeHTMLTags()
        .replaceHTMLEscapes()
        .replaceNumbers()
        .replaceNonAlphaNumericUnderscore()
        .collapseHyphens()
        .collapseUnderscores()
        .wsclean())

    def toListFromShingles(n: Int, ns: Int*): List[String] = (List(n) ++ ns.toList).flatMap{ i: Int => input.sliding(i) }.toList

    def toSequenceFromShingles(n: Int, ns: Int*): TextSequence = new TextSequence(toListFromShingles(n, ns: _*))

    def toList(sep: String = " "): List[String] = input.split(sep).toList

    def toSequence(sep: String = " "): TextSequence = new TextSequence(toList(sep))

    def isEmpty(): Boolean = input.isEmpty()
}
